internal/core/app_deploy.go

```diff
@@ -313,14 +315,48 @@ func (op *deployOperation) Do(ctx context.Context, log hclog.Logger, app *App, m
 		return nil, err
 	}
 
+	// NOTE(izaak): We to this to satisfy argmamper - because the Deploy function asks for this,
+	// we need to prove to argmapper that we can supply one. We'll swap this out
+	// with a different pointer on the plugin side. This one doesn't end up being used for anything.
+	declaredResourcesResp := &component.DeclaredResourcesResp{}
+
+	var args []argmapper.Arg
+	args = append(args, argmapper.Typed(declaredResourcesResp))
+	args = append(args, op.args()...)
+
 	val, err := app.callDynamicFunc(ctx,
 		log,
 		(*component.Deployment)(nil),
 		op.component,
 		op.component.Value.(component.Platform).DeployFunc(),
-		op.args()...,
+		args...,
 	)
 
+	// NOTE(izaak): Unfortunately, the returned type here is an waypoint-plugin-sdk/internal/plugioncomponent/Deployment,
+	// and because it's internal we can't cast to it. This is similar to how we're getting TemplateData
+	// (https://github.com/hashicorp/waypoint-plugin-sdk/blob/eda7ae600c2d69ba042ecac4d6d9333d37f6fd9d/docs/func.go#L106),
+	// but it still doesn't feel great.
```

Comment:
Hrm, we can reference release, so we should be able to reference deploy:

https://github.com/hashicorp/waypoint/blob/6306d37c290f71dc0f5013845732b7afe534fd9c/internal/core/app_release.go#L217-L228

I haven't messed with this myself so there might be something I'm missing.

Reply:
I think the difference is that for release, we explicitly cast to the super broad component.Release on the plugin side:

https://github.com/hashicorp/waypoint-plugin-sdk/blob/main/internal/plugin/releaser.go#L288

We may be able to do something similar here though to avoid the reflection - i'll give it a shot.

Reply:
Solved this in a different way. I'm now passing an outparameter into the plugin framework with argmapper, and extracting the goodies from the private struct in the plugin code (running as part of the core process) and passing it back into the outparameter:

https://github.com/hashicorp/waypoint-plugin-sdk/pull/40/files#diff-1131f9061dccc1444fff98f5fc696b66ab7180c942cd0b66d768f13822614fb1R302

So we're using argmapper outparameters in two different ways as part of this change!